File: android/app/src/main/java/com/example/jamboree/games/bluffmarket/BluffMarketServer.kt
package com.example.ubapp.games.bluffmarket

import android.content.Context
import com.example.ubapp.join.LoopbackGuest
import com.example.ubapp.social.GuestId
import com.example.ubapp.social.HostServer
import com.example.ubapp.stats.SeriesScore
import com.example.ubapp.stats.StatsStore
import com.example.ubapp.tutorials.GameTutorials
import org.json.JSONArray
import org.json.JSONObject

class BluffMarketServer(context: Context, val hostName: String = "Host") {
    val engine = BluffMarketEngine()
    private val server = HostServer(html = HostServer.htmlAsset(context, "bluff_market_browser.html"), ctx = context)
    private val appCtx = context.applicationContext
    private val guestToPlayer = HashMap<GuestId, String>()
    private val playerToGuest = HashMap<String, GuestId>()
    var onStateChange: (() -> Unit)? = null
    private var statRecorded = false
    private val series = SeriesScore()

    init { server.onMessage = ::onMessage; server.onLeave = ::onLeave }
    companion object { const val HOST_ID = "host" }

    fun start(): String? {
        engine.addPlayer(HOST_ID, hostName, isHost = true)
        val u = server.startServer()
        val local = server.attachLocalGuest()
        guestToPlayer[local] = HOST_ID
        playerToGuest[HOST_ID] = local
        emit()
        return u
    }

    fun makeLoopback(): LoopbackGuest = LoopbackGuest(server)
    fun stop() = server.stopServer()
    val guestCount: Int get() = server.guestCount

    fun hostSetOptions(o: BluffMarketOptions) {
        engine.setOptions(o); broadcastOptions(); emit()
    }
    fun hostStart() { engine.start(); broadcastState(); sendHandsPrivately(); emit() }
    fun hostNewGame() {
        engine.reset(); statRecorded = false
        broadcast(JSONObject().put("type", "reset")); broadcastLobby(); broadcastTutorialState()
        if (!series.isEmpty) broadcastSeries()
        emit()
    }
    fun hostFinalize() { engine.finalize(); broadcastState(); broadcastOver(); emit() }
    fun hostCallTutorialVote() = openTutorialVote()
    fun hostTutorialVote(yes: Boolean) = submitTutorialVote(HOST_ID, yes)
    fun hostDismissTutorial() { engine.tutorialVote.markShown(); broadcastTutorialState(); emit() }

    private fun onMessage(guest: GuestId, raw: String) {
        val j = runCatching { JSONObject(raw) }.getOrNull() ?: return
        val pid = guestToPlayer[guest]
        when (j.optString("type")) {
            "join" -> handleJoin(guest, j)
            "buy" -> pid?.let { applyResult(it, engine.buyFromMarket(it)) }
            "sell" -> pid?.let { applyResult(it, engine.sellToMarket(it, j.optString("cardId"))) }
            "propose_trade" -> pid?.let {
                applyResult(it, engine.proposeTrade(it, j.optString("targetId"), j.optString("cardId")))
            }
            "counter_trade" -> pid?.let {
                applyResult(it, engine.counterTrade(it, j.optString("cardId")))
            }
            "decline_trade" -> pid?.let { applyResult(it, engine.declineTrade(it)) }
            "guarantee" -> pid?.let { applyResult(it, engine.useGuarantee(it)) }
            "respond_trade" -> pid?.let {
                applyResult(it, engine.respondTrade(it, j.optBoolean("accept")))
            }
            "finalize" -> if (engine.phase == BluffMarketPhase.SCORING) hostFinalize()
            "set_options" -> Unit
            "call_tutorial_vote" -> openTutorialVote()
            "tutorial_vote" -> pid?.let { submitTutorialVote(it, j.getBoolean("yes")) }
        }
    }

    private fun applyResult(pid: String, err: String?) {
        if (err == null) {
            broadcastState(); sendHandsPrivately()
            if (engine.phase == BluffMarketPhase.SCORING) broadcastScores()
            if (engine.phase == BluffMarketPhase.GAME_OVER) broadcastOver()
            emit()
        } else playerToGuest[pid]?.let { send(it, JSONObject().put("type", "error").put("message", err)) }
    }

    private fun onLeave(guest: GuestId) {
        val pid = guestToPlayer.remove(guest) ?: return
        playerToGuest.remove(pid)
        if (engine.phase == BluffMarketPhase.LOBBY) {
            engine.removePlayer(pid); engine.tutorialVote.removeVoter(pid); broadcastLobby()
            if (engine.tutorialVote.isOpen || engine.tutorialVote.hasResult) broadcastTutorialState()
        }
        emit()
    }

    private fun handleJoin(guest: GuestId, j: JSONObject) {
        if (engine.phase != BluffMarketPhase.LOBBY) {
            send(guest, JSONObject().put("type", "error").put("message", "Game already in progress")); return
        }
        val name = j.optString("name").trim(); if (name.isEmpty()) return
        val pid = "g${guestToPlayer.size + 1}"
        engine.addPlayer(pid, name)
        guestToPlayer[guest] = pid; playerToGuest[pid] = guest
        send(guest, JSONObject().put("type", "welcome").put("yourId", pid).put("yourName", name).put("game", "bluff_market"))
        broadcastLobby(); broadcastOptions(); broadcastTutorialState()
        if (!series.isEmpty) broadcastSeries()
        emit()
    }

    private fun broadcastOptions() {
        broadcast(JSONObject().put("type", "options")
            .put("turnsPerPlayer", engine.options.turnsPerPlayer)
            .put("twoBombs", engine.options.twoBombs)
            .put("wildcard", engine.options.wildcard))
    }

    private fun broadcastLobby() {
        val arr = JSONArray()
        for (p in engine.players.values) arr.put(JSONObject().put("id", p.id).put("name", p.name).put("isHost", p.isHost))
        broadcast(JSONObject().put("type", "lobby").put("players", arr))
    }

    private fun broadcastState() {
        val playersArr = JSONArray()
        for (p in engine.players.values) {
            playersArr.put(JSONObject()
                .put("id", p.id).put("name", p.name)
                .put("handCount", p.hand.size)
                .put("coins", p.coins)
                .put("turnsTaken", p.turnsTaken)
                .put("guaranteeUsed", p.guaranteeUsed))
        }
        val phase = when (engine.phase) {
            BluffMarketPhase.LOBBY -> "lobby"
            BluffMarketPhase.PLAYING -> "playing"
            BluffMarketPhase.SCORING -> "scoring"
            BluffMarketPhase.GAME_OVER -> "gameOver"
        }
        val payload = JSONObject()
            .put("type", "state")
            .put("phase", phase)
            .put("currentId", engine.current?.id ?: JSONObject.NULL)
            .put("marketSize", engine.market.size)
            .put("lastEvent", engine.lastEvent ?: "")
            .put("players", playersArr)
            .put("turnsPerPlayer", engine.options.turnsPerPlayer)
        engine.activeTrade?.let { t ->
            val tjson = JSONObject()
                .put("proposerId", t.proposerId)
                .put("targetId", t.targetId)
                .put("proposerCommitted", t.proposerCardId != null)
                .put("targetCommitted", t.targetCardId != null)
                .put("revealed", t.revealed)
                .put("proposerGuarantee", t.proposerGuarantee)
                .put("targetGuarantee", t.targetGuarantee)
            t.proposerAccept?.let { tjson.put("proposerAccept", it) }
            t.targetAccept?.let { tjson.put("targetAccept", it) }
            if (t.revealed) {
                t.proposerCardId?.let { engine.cardCatalog[it]?.let { c -> tjson.put("proposerCard", cardJson(c)) } }
                t.targetCardId?.let { engine.cardCatalog[it]?.let { c -> tjson.put("targetCard", cardJson(c)) } }
            }
            payload.put("trade", tjson)
        }
        broadcast(payload)
    }

    private fun cardJson(c: BluffCard): JSONObject {
        val (kind, value) = when (val k = c.kind) {
            is BluffKind.Points -> "points" to k.value
            is BluffKind.Bomb -> "bomb" to k.value
            is BluffKind.Wildcard -> "wildcard" to 0
        }
        return JSONObject()
            .put("id", c.id).put("kind", kind)
            .put("value", value).put("label", c.label)
    }

    private fun sendHandsPrivately() {
        for (p in engine.players.values) {
            val g = playerToGuest[p.id] ?: continue
            val cards = JSONArray()
            for (c in p.hand) cards.put(cardJson(c))
            send(g, JSONObject().put("type", "hand").put("cards", cards))
        }
    }

    private fun broadcastScores() {
        val arr = JSONArray()
        for (r in engine.score()) {
            arr.put(JSONObject()
                .put("id", r.id).put("name", r.name)
                .put("total", r.total).put("sum", r.sum)
                .put("coins", r.coins).put("hasBomb", r.hasBomb))
        }
        broadcast(JSONObject().put("type", "scores").put("rows", arr))
    }

    private fun broadcastOver() {
        val rows = engine.score()
        val winner = rows.maxByOrNull { it.total }
        if (!statRecorded) {
            statRecorded = true
            val names = ArrayList<String>()
            winner?.let { names.add(it.name) }
            for (r in rows) if (r.id != winner?.id) names.add(r.name)
            StatsStore.record(appCtx, "bluff_market", names, "win")
            winner?.let { series.record(it.name); broadcastSeries() }
        }
        val arr = JSONArray()
        for (r in rows) {
            arr.put(JSONObject()
                .put("id", r.id).put("name", r.name)
                .put("total", r.total).put("sum", r.sum)
                .put("coins", r.coins).put("hasBomb", r.hasBomb))
        }
        broadcast(JSONObject().put("type", "over")
            .put("winnerId", winner?.id ?: JSONObject.NULL)
            .put("rows", arr))
    }

    private fun openTutorialVote() {
        if (engine.phase != BluffMarketPhase.LOBBY) return
        if (engine.tutorialVote.isOpen || engine.tutorialVote.tutorialShown) return
        engine.tutorialVote.open(engine.players.keys); broadcastTutorialState(); emit()
    }
    private fun submitTutorialVote(voterId: String, yes: Boolean) {
        if (!engine.tutorialVote.isOpen) return
        engine.tutorialVote.submit(voterId, yes); broadcastTutorialState(); emit()
    }
    private fun broadcastTutorialState() {
        val v = engine.tutorialVote
        val p = JSONObject().put("type", "tutorial_vote_state")
            .put("isOpen", v.isOpen).put("yesCount", v.yesCount).put("noCount", v.noCount)
            .put("eligibleCount", v.eligibleCount).put("result", v.result ?: JSONObject.NULL)
            .put("tutorialShown", v.tutorialShown)
        if (v.result == true && !v.tutorialShown) {
            p.put("title", GameTutorials.bluffMarket.title)
            p.put("sections", JSONArray(GameTutorials.bluffMarket.sectionsJson().map { JSONObject(it as Map<*, *>) }))
            p.put("menuSections", JSONArray(GameTutorials.bluffMarket.browserMenuSectionsJson().map { JSONObject(it as Map<*, *>) }))
        }
        broadcast(p)
    }

    private fun broadcastSeries() {
        val scores = JSONObject()
        for ((k, v) in series.scores) scores.put(k, v)
        broadcast(JSONObject().put("type", "series_state").put("rounds", series.rounds).put("scores", scores))
    }

    private fun emit() { onStateChange?.invoke() }
    private fun broadcast(o: JSONObject) = server.broadcast(o.toString())
    private fun send(g: GuestId, o: JSONObject) = server.send(g, o.toString())
}

File: ios/Jamboree/Games/BluffMarket/BluffMarketServer.swift
import Foundation

/// Wraps [HostServer] with Bluff Market routing.
final class BluffMarketServer {
    private(set) var engine = BluffMarketEngine()
    static let hostId = "host"
    let hostName: String

    private let server: HostServer
    private var guestToPlayer: [GuestId: String] = [:]
    private var playerToGuest: [String: GuestId] = [:]

    var onStateChange: (() -> Void)?
    private var statRecorded = false
    private let series = SeriesScore()

    init(server: HostServer? = nil, hostName: String = "Host") {
        self.server = server ?? HostServer(html: HostServer.htmlResource(named: "bluff_market_browser"))
        self.hostName = hostName
        self.server.onMessage = { [weak self] id, raw in self?.onMessage(from: id, raw: raw) }
        self.server.onLeave = { [weak self] id in self?.onLeave(id) }
    }

    @discardableResult
    func start() throws -> URL? {
        engine.addPlayer(id: Self.hostId, name: hostName, isHost: true)
        let url = try server.start()
        let local = server.attachLocalGuest()
        guestToPlayer[local] = Self.hostId
        playerToGuest[Self.hostId] = local
        emit()
        return url
    }

    @MainActor
    func makeLoopback() -> LoopbackGuest { LoopbackGuest(server: server) }
    func stop() { server.stop(); resetState() }

    /// Clear all per-session state so the next time the host starts
    /// hosting they get a fresh screen — empty lobby, tutorial vote
    /// available again.
    private func resetState() {
        engine = BluffMarketEngine()
        guestToPlayer.removeAll(); playerToGuest.removeAll()
        statRecorded = false
        emit()
    }
    var guestCount: Int { server.guestCount }

    func hostSetOptions(_ o: BluffMarketOptions) {
        engine.setOptions(o); broadcastOptions(); emit()
    }
    func hostStart() {
        engine.start()
        broadcastState(); sendHandsPrivately(); emit()
    }
    func hostNewGame() {
        engine.reset(); statRecorded = false
        broadcast(["type": "reset"]); broadcastLobby(); broadcastTutorialState()
        if !series.isEmpty { broadcastSeries() }
        emit()
    }
    func hostFinalize() {
        engine.finalize()
        broadcastState(); broadcastOver(); emit()
    }
    func hostCallTutorialVote() { openTutorialVote() }
    func hostTutorialVote(_ yes: Bool) { submitTutorialVote(voterId: Self.hostId, yes: yes) }
    func hostDismissTutorial() {
        engine.tutorialVote.markShown(); broadcastTutorialState(); emit()
    }

    private func onMessage(from guest: GuestId, raw: String) {
        guard let data = raw.data(using: .utf8),
              let j = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let type = j["type"] as? String else { return }
        switch type {
        case "join": handleJoin(guest, json: j)
        case "buy":
            if let pid = guestToPlayer[guest] { applyResult(pid, engine.buyFromMarket(playerId: pid)) }
        case "sell":
            if let pid = guestToPlayer[guest], let cid = j["cardId"] as? String {
                applyResult(pid, engine.sellToMarket(playerId: pid, cardId: cid))
            }
        case "propose_trade":
            if let pid = guestToPlayer[guest],
               let target = j["targetId"] as? String,
               let cid = j["cardId"] as? String {
                applyResult(pid, engine.proposeTrade(playerId: pid, targetId: target, cardId: cid))
            }
        case "counter_trade":
            if let pid = guestToPlayer[guest], let cid = j["cardId"] as? String {
                applyResult(pid, engine.counterTrade(playerId: pid, cardId: cid))
            }
        case "decline_trade":
            if let pid = guestToPlayer[guest] { applyResult(pid, engine.declineTrade(playerId: pid)) }
        case "guarantee":
            if let pid = guestToPlayer[guest] { applyResult(pid, engine.useGuarantee(playerId: pid)) }
        case "respond_trade":
            if let pid = guestToPlayer[guest], let accept = j["accept"] as? Bool {
                applyResult(pid, engine.respondTrade(playerId: pid, accept: accept))
            }
        case "finalize":
            // Anyone can request final scoring once phase == scoring, host usually drives it.
            if engine.phase == .scoring { hostFinalize() }
        case "set_options":
            break
        case "call_tutorial_vote": if guestToPlayer[guest] != nil { openTutorialVote() }
        case "tutorial_vote":
            if let pid = guestToPlayer[guest], let yes = j["yes"] as? Bool {
                submitTutorialVote(voterId: pid, yes: yes)
            }
        default: break
        }
    }

    private func applyResult(_ pid: String, _ err: String?) {
        if err == nil {
            broadcastState(); sendHandsPrivately()
            if engine.phase == .scoring { broadcastScores() }
            if engine.phase == .gameOver { broadcastOver() }
            emit()
        } else if let g = playerToGuest[pid] {
            send(g, ["type": "error", "message": err!])
        }
    }

    private func onLeave(_ guest: GuestId) {
        guard let pid = guestToPlayer.removeValue(forKey: guest) else { return }
        playerToGuest[pid] = nil
        if engine.phase == .lobby {
            engine.removePlayer(pid)
            engine.tutorialVote.removeVoter(pid)
            broadcastLobby()
            if engine.tutorialVote.isOpen || engine.tutorialVote.hasResult {
                broadcastTutorialState()
            }
        }
        emit()
    }

    private func handleJoin(_ guest: GuestId, json: [String: Any]) {
        if engine.phase != .lobby {
            send(guest, ["type": "error", "message": "Game already in progress"])
            return
        }
        let name = String(((json["name"] as? String) ?? "").trimmingCharacters(in: .whitespaces).prefix(24))
        guard !name.isEmpty else { return }
        let pid = "p\(guest.value)"
        engine.addPlayer(id: pid, name: name)
        guestToPlayer[guest] = pid
        playerToGuest[pid] = guest
        send(guest, ["type": "welcome", "yourId": pid, "yourName": name, "game": "bluff_market"])
        broadcastLobby(); broadcastOptions(); broadcastTutorialState()
        if !series.isEmpty { broadcastSeries() }
        emit()
    }

    private func broadcastOptions() {
        broadcast([
            "type": "options",
            "turnsPerPlayer": engine.options.turnsPerPlayer,
            "twoBombs": engine.options.twoBombs,
            "wildcard": engine.options.wildcard,
        ])
    }

    private func broadcastLobby() {
        broadcast([
            "type": "lobby",
            "players": engine.players.values.map { ["id": $0.id, "name": $0.name, "isHost": $0.isHost] },
        ])
    }

    private func broadcastState() {
        let phase: String
        switch engine.phase {
        case .lobby: phase = "lobby"
        case .playing: phase = "playing"
        case .scoring: phase = "scoring"
        case .gameOver: phase = "gameOver"
        }
        var payload: [String: Any] = [
            "type": "state",
            "phase": phase,
            "currentId": engine.current?.id as Any,
            "marketSize": engine.market.count,
            "lastEvent": engine.lastEvent ?? "",
            "players": engine.players.values.map {
                ["id": $0.id, "name": $0.name, "handCount": $0.hand.count,
                 "coins": $0.coins, "turnsTaken": $0.turnsTaken,
                 "guaranteeUsed": $0.guaranteeUsed] as [String: Any]
            },
            "turnsPerPlayer": engine.options.turnsPerPlayer,
        ]
        if let t = engine.activeTrade {
            // Public: who's in the trade and (only once both committed) the cards.
            var tradeJson: [String: Any] = [
                "proposerId": t.proposerId,
                "targetId": t.targetId,
                "proposerCommitted": t.proposerCardId != nil,
                "targetCommitted": t.targetCardId != nil,
                "revealed": t.revealed,
                "proposerGuarantee": t.proposerGuarantee,
                "targetGuarantee": t.targetGuarantee,
            ]
            if t.proposerAccept != nil { tradeJson["proposerAccept"] = t.proposerAccept! }
            if t.targetAccept != nil { tradeJson["targetAccept"] = t.targetAccept! }
            if t.revealed,
               let pcid = t.proposerCardId, let pc = engine.cardCatalog[pcid],
               let tcid = t.targetCardId, let tc = engine.cardCatalog[tcid] {
                tradeJson["proposerCard"] = cardJson(pc)
                tradeJson["targetCard"] = cardJson(tc)
            }
            payload["trade"] = tradeJson
        }
        broadcast(payload)
    }

    private func cardJson(_ c: BluffCard) -> [String: Any] {
        var kind = "points"; var v: Int = c.points
        switch c.kind {
        case .points: kind = "points"
        case .bomb: kind = "bomb"
        case .wildcard: kind = "wildcard"; v = 0
        }
        return ["id": c.id, "kind": kind, "value": v, "label": c.label]
    }

    private func sendHandsPrivately() {
        for p in engine.players.values {
            guard let g = playerToGuest[p.id] else { continue }
            send(g, [
                "type": "hand",
                "cards": p.hand.map { cardJson($0) },
            ])
        }
    }

    private func broadcastScores() {
        let scores = engine.score()
        broadcast([
            "type": "scores",
            "rows": scores.map { row in
                [
                    "id": row.id, "name": row.name,
                    "total": row.total, "sum": row.sum, "coins": row.coins,
                    "hasBomb": row.hasBomb,
                ] as [String: Any]
            },
        ])
    }

    private func broadcastOver() {
        let scores = engine.score()
        let winner = scores.max(by: { $0.total < $1.total })
        if !statRecorded {
            statRecorded = true
            var names: [String] = []
            if let w = winner { names.append(w.name) }
            for r in scores where r.id != winner?.id { names.append(r.name) }
            StatsStore.record(gameId: "bluff_market", players: names, outcome: "win")
            if let w = winner { series.record(w.name); broadcastSeries() }
        }
        broadcast([
            "type": "over",
            "winnerId": winner?.id as Any,
            "rows": scores.map { row in
                [
                    "id": row.id, "name": row.name,
                    "total": row.total, "sum": row.sum, "coins": row.coins,
                    "hasBomb": row.hasBomb,
                ] as [String: Any]
            },
        ])
    }

    private func openTutorialVote() {
        guard engine.phase == .lobby, !engine.tutorialVote.isOpen, !engine.tutorialVote.tutorialShown else { return }
        engine.tutorialVote.open(eligibleIds: engine.players.keys)
        broadcastTutorialState(); emit()
    }
    private func submitTutorialVote(voterId: String, yes: Bool) {
        guard engine.tutorialVote.isOpen else { return }
        engine.tutorialVote.submit(voterId: voterId, yes: yes)
        broadcastTutorialState(); emit()
    }
    private func broadcastTutorialState() {
        let v = engine.tutorialVote
        var payload: [String: Any] = [
            "type": "tutorial_vote_state",
            "isOpen": v.isOpen, "yesCount": v.yesCount, "noCount": v.noCount,
            "eligibleCount": v.eligibleCount, "result": v.result as Any, "tutorialShown": v.tutorialShown,
        ]
        if v.result == true && !v.tutorialShown {
            payload["title"] = GameTutorials.bluffMarket.title
            payload["sections"] = GameTutorials.bluffMarket.sectionsJSON()
            payload["menuSections"] = GameTutorials.bluffMarket.browserMenuSectionsJSON()
        }
        broadcast(payload)
    }

    private func broadcastSeries() {
        var scores: [String: Int] = [:]
        for entry in series.scores { scores[entry.key] = entry.value }
        broadcast(["type": "series_state", "rounds": series.rounds, "scores": scores])
    }

    private func emit() { onStateChange?() }
    private func broadcast(_ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let s = String(data: data, encoding: .utf8) else { return }
        server.broadcast(s)
    }
    private func send(_ guest: GuestId, _ payload: [String: Any]) {
        guard let data = try? JSONSerialization.data(withJSONObject: payload),
              let s = String(data: data, encoding: .utf8) else { return }
        server.send(to: guest, s)
    }
}
